Sources/TuistCore/Graph/GraphTraverser.swift

```diff
@@ -593,6 +612,61 @@ public class GraphTraverser: GraphTraversing {
         return references
     }
 
+    /// Recursively find platform filters within transitive dependencies
+    /// - Parameters:
+    ///   - rootDependency: dependency whose platform filters we need when depending on `transitiveDependency`
+    ///   - transitiveDependency: target dependency
+    /// - Returns: PlatformFilters to apply to transitive dependency or `nil` if the path to a dependency results in a disjoint
+    /// set of platform filters.
+    func platformFilters(from rootDependency: GraphDependency, to transitiveDependency: GraphDependency) -> PlatformFilters? {
+        var visited: Set<GraphDependency> = []
+
+        func find(from root: GraphDependency, to other: GraphDependency) -> PlatformFilters? {
+            guard !visited.contains(root) else { return nil }
+            visited.insert(root)
+            guard let dependencies = graph.dependencies[root] else { return nil }
+
+            if dependencies.contains(other) {
+                // If we reach the end and there's no filters,
+                // return empty to signify the dependency has no filters
+                return graph.dependencyPlatformFilters[(root, other)] ?? []
+            } else {
+                let filters = dependencies.map { node -> PlatformFilters? in
+
+                    // If an intervening dependency has filters, we need to constrain downstream filters to a subset of those.
+                    if let currentDependencyPlatformFilters = graph.dependencyPlatformFilters[(root, node)] {
+                        guard let transitive = find(from: node, to: other) else {
+                            return nil
+                        }
+
+                        // If there are transitive filters, return the intersection with the current filters
+                        // If the intersection is empty, return `nil` to signify the dependency should be trimmed.
+                        if !transitive.isEmpty {
+                            let intersection = transitive.intersection(currentDependencyPlatformFilters)
+                            return intersection.isEmpty ? nil : intersection
+                        } else {
+                            return currentDependencyPlatformFilters
+                        }
+                    } else { // Otherwise, just find the filters for this
+                        return find(from: node, to: other)
+                    }
+                }
+
+                let transitiveFilters = filters.reduce(Set<PlatformFilter>?(nil)) { result, otherFilters in
+                    if let result {
+                        return result.union(otherFilters ?? [])
+                    } else {
+                        return otherFilters
+                    }
```

Comment:
Does the logic here need to give more precedence to no filters ?

For example:

![Screenshot 2023-11-15 at 09 07 44](https://github.com/tuist/tuist/assets/11914919/c98bbff3-62dc-416f-aa28-086c8fcdbcbd)

When asking for the platform filters for `App > Static Framework C` it should return no filters as opposed to `.macOS`


```swift
    //
    //      -> Static Framework A (platformFilter = [.mac]) ->
    //  App                                                    Static Framewowork C
    //      -> Static Framework B                           ->
    //
    func test_platformFilters_transitivePlatformFilter_noFiltersHaveHigherPrecedence() throws {
        // Given
        let app = Target.test(name: "App", destinations: [.iPad, .iPhone, .mac], product: .app)
        let staticFrameworkA = Target.test(
            name: "StaticFrameworkA",
            destinations: [.iPad, .iPhone, .mac],
            product: .staticLibrary
        )
        
        let staticFrameworkB = Target.test(
            name: "StaticFrameworkB",
            destinations: [.mac],
            product: .staticLibrary
        )
        
        let staticFrameworkC = Target.test(
            name: "StaticFrameworkC",
            destinations: [.iPad, .iPhone, .mac],
            product: .staticLibrary
        )
        
        let project = Project.test(targets: [
            app,
            staticFrameworkA,
            staticFrameworkB,
            staticFrameworkC,
        ])
        
        let appkGraphDependency: GraphDependency = .target(name: app.name, path: project.path)
        let staticFrameworkAGraphDependency: GraphDependency = .target(name: staticFrameworkA.name, path: project.path)
        let staticFrameworkBGraphDependency: GraphDependency = .target(name: staticFrameworkB.name, path: project.path)
        let staticFrameworkCGraphDependency: GraphDependency = .target(name: staticFrameworkC.name, path: project.path)
        
        let graph = Graph.test(
            path: project.path,
            projects: [project.path: project],
            targets: [
                project.path: [
                    app.name: app,
                    staticFrameworkA.name: staticFrameworkA,
                    staticFrameworkB.name: staticFrameworkB,
                    staticFrameworkC.name: staticFrameworkC,
                ]
            ],
            dependencies: [
                appkGraphDependency: [
                    staticFrameworkAGraphDependency,
                    staticFrameworkBGraphDependency,
                ],
                staticFrameworkAGraphDependency: [
                    staticFrameworkCGraphDependency,
                ],
                staticFrameworkBGraphDependency: [
                    staticFrameworkCGraphDependency,
                ]
            ],
            edges: [
                GraphEdge(from: appkGraphDependency, to: staticFrameworkBGraphDependency): [.macos],
            ]
        )
        let subject = GraphTraverser(graph: graph)
        
        // When
        let results = subject.platformFilters(from: appkGraphDependency, to: staticFrameworkCGraphDependency)
        
        // Then
        XCTAssertEqual(results?.sorted(), [])
    }
```